File: answers/lintcode/Distinct Subsequences.cpp
class Solution {
public:
    /**
     * @param S, T: Two string.
     * @return: Count the number of distinct subsequences
     */
    int numDistinct(string &S, string &T) {
        // write your code here
        vector<vector<int> > dp(S.length(), vector<int> (T.length(), 0));
        for (size_t i = 0; i < S.length(); ++ i) {
            for (size_t j = 0; j < T.length(); ++ j) {
                if (S[i] == T[j]) {
                    dp[i][j] += (i > 0 && j > 0) ? dp[i - 1][j - 1] : ((i > 0 || (0 == i && 0 == j)) ? 1 : 0);
                }
                if (i > 0) {
                    dp[i][j] += dp[i - 1][j];
                }
            }
        }
        return dp.back().back();
    }
};

File: answers/leetcode/Distinct Subsequences/Distinct Subsequences.java
class Solution {
    public int numDistinct(String s, String t) {
        int[][] dp = new int[s.length() + 1][t.length() + 1];
        for (int i = 0; i <= s.length(); i++) {
            for (int j = 0; j <= t.length(); j++) {
                int count = 0;
                if (i == 0 && j == 0) {
                    count++;
                } else if (i > 0) {
                    if (j > 0 && s.charAt(i - 1) == t.charAt(j - 1)) {
                        count += dp[i - 1][j - 1];
                    }
                    count += dp[i - 1][j];
                }
                dp[i][j] = count;
            }
        }
        return dp[s.length()][t.length()];
    }
}